File: backend/app/schemas/schedule.py
"""
Схемы для расписания и занятий.
"""

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.schedule import DayOfWeek, LessonType, WeekParity
from app.schemas.schedule_parser import ScheduleConflictResponse

AttendanceSummaryState = Literal["hidden", "not_applicable", "unmarked", "partial", "complete"]

# === ScheduleItem ===


class ScheduleItemBase(BaseModel):
    day_of_week: DayOfWeek
    lesson_number: int = Field(ge=1, le=8)
    lesson_type: LessonType
    subject: str | None = None
    subject_id: UUID | None = None
    offering_id: UUID | None = None
    room: str | None = None
    teacher_id: UUID | None = None
    start_date: date
    end_date: date | None = None
    week_parity: WeekParity | None = None
    subgroup: int | None = Field(None, ge=1, le=2)


class ScheduleItemCreate(ScheduleItemBase):
    group_id: UUID


class ScheduleItemUpdate(BaseModel):
    day_of_week: DayOfWeek | None = None
    lesson_number: int | None = Field(None, ge=1, le=8)
    lesson_type: LessonType | None = None
    subject: str | None = None
    subject_id: UUID | None = None
    offering_id: UUID | None = None
    room: str | None = None
    teacher_id: UUID | None = None
    start_date: date | None = None
    end_date: date | None = None
    week_parity: WeekParity | None = None
    subgroup: int | None = Field(None, ge=1, le=2)
    is_active: bool | None = None


class ScheduleItemResponse(ScheduleItemBase):
    id: UUID
    group_id: UUID
    is_active: bool

    model_config = ConfigDict(from_attributes=True)


# === Lesson ===


class LessonBase(BaseModel):
    date: date
    lesson_number: int = Field(ge=1, le=8)
    lesson_type: LessonType
    subject_id: UUID | None = None
    offering_id: UUID | None = None
    topic: str | None = None
    room: str | None = None
    work_id: UUID | None = None
    subgroup: int | None = Field(None, ge=1, le=2)


class LessonCreate(LessonBase):
    group_id: UUID
    schedule_item_id: UUID | None = None


class LessonUpdate(BaseModel):
    subject_id: UUID | None = None
    offering_id: UUID | None = None
    topic: str | None = None
    work_id: UUID | None = None
    work_number: int | None = Field(None, ge=1, le=20, description="Номер лабы/практики")
    is_cancelled: bool | None = None
    cancellation_reason: str | None = None
    ended_early: bool | None = None


class LessonAttendanceSummaryResponse(BaseModel):
    state: AttendanceSummaryState
    marked_count: int = 0
    expected_count: int = 0
    is_past: bool


class LessonResponse(LessonBase):
    id: UUID
    group_id: UUID
    schedule_item_id: UUID | None
    is_cancelled: bool
    cancellation_reason: str | None
    work_number: int | None = None
    ended_early: bool = False
    subject_name: str | None = None
    group_name: str | None = None
    summary: LessonAttendanceSummaryResponse | None = None

    model_config = ConfigDict(from_attributes=True)


class ScheduleAttendanceUpdate(BaseModel):
    student_id: UUID
    status: str | None = None


class ScheduleGradeUpdate(BaseModel):
    student_id: UUID
    grade: int | None = Field(None, ge=2, le=5)
    work_number: int | None = Field(None, ge=1, le=20, description="Номер сдаваемой работы")


class LessonSheetSaveRequest(BaseModel):
    topic: str | None = None
    lesson_work_number: int | None = Field(None, ge=1, le=20, description="Базовый номер работы пары")
    status: Literal["normal", "cancelled", "early"]
    attendance_updates: list[ScheduleAttendanceUpdate] = Field(default_factory=list)
    grade_updates: list[ScheduleGradeUpdate] = Field(default_factory=list)


class GroupedLectureSheetSaveItem(BaseModel):
    lesson_id: UUID
    attendance_updates: list[ScheduleAttendanceUpdate] = Field(default_factory=list)


class GroupedLectureSheetSaveRequest(BaseModel):
    status: Literal["normal", "cancelled", "early"]
    topic: str | None = None
    items: list[GroupedLectureSheetSaveItem] = Field(default_factory=list)


class LessonSheetAttendanceResponse(BaseModel):
    student_id: UUID
    status: str


class LessonSheetGradeResponse(BaseModel):
    student_id: UUID
    grade: int | None = None
    work_number: int | None = None
    has_conflict: bool = False
    conflict_count: int = 0
    grade_items: list[dict[str, int | None]] = Field(default_factory=list)


class LessonSheetSaveResponse(BaseModel):
    lesson: LessonResponse
    attendance: list[LessonSheetAttendanceResponse]
    grades: list[LessonSheetGradeResponse]


class GroupedLectureSheetItemResponse(BaseModel):
    lesson_id: UUID
    attendance: list[LessonSheetAttendanceResponse]


class GroupedLectureSheetSaveResponse(BaseModel):
    items: list[GroupedLectureSheetItemResponse]


class GroupedLectureGroupResponse(BaseModel):
    id: UUID
    name: str
    lesson_id: UUID


class GroupedLectureAttendanceSummaryResponse(BaseModel):
    state: AttendanceSummaryState
    marked_count: int = 0
    expected_count: int = 0
    is_past: bool


class GroupedLectureResponse(BaseModel):
    date: date
    lesson_number: int
    subject_id: UUID | None = None
    subject_name: str | None = None
    topic: str | None = None
    room: str | None = None
    is_cancelled: bool = False
    ended_early: bool = False
    groups: list[GroupedLectureGroupResponse]
    summary: GroupedLectureAttendanceSummaryResponse | None = None


class ScheduleParseStatusResponse(BaseModel):
    is_running: bool
    status: str | None = None
    started_at: str | None = None
    finished_at: str | None = None
    lessons_created: int | None = None
    lessons_updated: int | None = None
    lessons_skipped: int | None = None
    conflicts_created: int | None = None
    error_message: str | None = None
    last_run: str | None = None


class ScheduleViewResponse(BaseModel):
    """Aggregate payload for admin schedule page."""

    parse_status: ScheduleParseStatusResponse
    conflicts: list[ScheduleConflictResponse]
    lessons: list[LessonResponse]
    grouped_lectures: list[GroupedLectureResponse]
    last_updated: str


# === Bulk operations ===


class GenerateLessonsRequest(BaseModel):
    """Запрос на генерацию занятий из расписания"""

    group_id: UUID
    start_date: date
    end_date: date


class GenerateLessonsResponse(BaseModel):
    """Ответ на генерацию занятий"""

    created_count: int
    lessons: list[LessonResponse]


class ParseScheduleRequest(BaseModel):
    """Запрос на парсинг расписания"""

    teacher_name: str
    start_date: date
    end_date: date | None = None  # По умолчанию - сегодня


class ParseScheduleResponse(BaseModel):
    """Результат парсинга"""

    total_parsed: int
    groups_created: int
    lessons_created: int
    lessons_updated: int = 0
    lessons_skipped: int
    conflicts_created: int = 0
    subjects_created: int = 0
    assignments_created: int = 0
    groups: list[str]
    subjects: list[str] = []
    semester_end_detected: bool = False
    last_lesson_date: str | None = None
    empty_weeks_count: int = 0

File: frontend/src/lib/api/types/schedule.ts
export type DayOfWeek = 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday';
export type LessonType = 'lecture' | 'practice' | 'lab';
export type WeekParity = 'odd' | 'even';
export type AttendanceSummaryState = 'hidden' | 'not_applicable' | 'unmarked' | 'partial' | 'complete';

export interface LessonAttendanceSummaryResponse {
  state: AttendanceSummaryState;
  marked_count: number;
  expected_count: number;
  is_past: boolean;
}

export interface GroupedLectureAttendanceSummaryResponse {
  state: AttendanceSummaryState;
  marked_count: number;
  expected_count: number;
  is_past: boolean;
}

export interface ScheduleItemCreate {
  day_of_week: DayOfWeek;
  lesson_number: number;
  lesson_type: LessonType;
  subject?: string;
  room?: string;
  teacher_id?: string;
  start_date: string;
  end_date?: string;
  week_parity?: WeekParity;
  subgroup?: number;
}

export interface ScheduleItemUpdate {
  day_of_week?: DayOfWeek;
  lesson_number?: number;
  lesson_type?: LessonType;
  subject?: string;
  room?: string;
  teacher_id?: string;
  start_date?: string;
  end_date?: string;
  week_parity?: WeekParity;
  subgroup?: number;
  is_active?: boolean;
}

export interface ScheduleItemResponse {
  id: string;
  group_id: string;
  day_of_week: DayOfWeek;
  lesson_number: number;
  lesson_type: LessonType;
  subject?: string;
  room?: string;
  teacher_id?: string;
  start_date: string;
  end_date?: string;
  week_parity?: WeekParity;
  subgroup?: number;
  is_active: boolean;
}

export interface LessonCreate {
  group_id: string;
  schedule_item_id?: string;
  date: string;
  lesson_number: number;
  lesson_type: LessonType;
  topic?: string;
  work_id?: string;
  subgroup?: number;
}

export interface LessonUpdate {
  topic?: string;
  work_id?: string;
  work_number?: number | null;
  is_cancelled?: boolean;
  cancellation_reason?: string;
}

export interface LessonResponse {
  id: string;
  group_id: string;
  schedule_item_id?: string;
  date: string;
  lesson_number: number;
  lesson_type: LessonType;
  topic?: string;
  work_id?: string;
  work_number?: number | null;
  subgroup?: number;
  is_cancelled: boolean;
  cancellation_reason?: string;
  ended_early: boolean;
  subject_name?: string | null;
  group_name?: string | null;
  summary?: LessonAttendanceSummaryResponse | null;
}

export interface GenerateLessonsResponse {
  created_count: number;
  lessons: LessonResponse[];
}

export interface ScheduleParserConfig {
  enabled: boolean;
  teacher_name: string;
  days_of_week: number[];
  run_time: string;
  parse_days_ahead: number;
}

export interface ScheduleParserConfigResponse extends ScheduleParserConfig {
  id: string;
  teacher_id: string;
  last_run_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ParseScheduleResponse {
  total_parsed: number;
  groups_created: number;
  lessons_created: number;
  lessons_updated: number;
  lessons_skipped: number;
  conflicts_created: number;
  subjects_created: number;
  assignments_created: number;
  groups: string[];
  subjects: string[];
  semester_end_detected: boolean;
  last_lesson_date: string | null;
  empty_weeks_count: number;
}

export type ScheduleAutoParseResponse = ParseScheduleResponse;

export interface ScheduleConflict {
  id: string;
  lesson_id: string;
  conflict_type: 'changed' | 'deleted';
  old_data: {
    topic?: string;
    lesson_type?: string;
    room?: string;
    date?: string;
    lesson_number?: number;
  };
  new_data: {
    topic?: string;
    lesson_type?: string;
    room?: string;
    date?: string;
    lesson_number?: number;
  } | null;
  created_at: string;
}

export interface GroupedLectureGroup {
  id: string;
  name: string;
  lesson_id: string;
}

export interface GroupedLectureResponse {
  date: string;
  lesson_number: number;
  subject_id: string | null;
  subject_name: string | null;
  topic: string | null;
  is_cancelled: boolean;
  ended_early: boolean;
  groups: GroupedLectureGroup[];
  summary?: GroupedLectureAttendanceSummaryResponse | null;
}

export interface ScheduleParseStatusResponse {
  is_running: boolean;
  status?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  lessons_created?: number | null;
  lessons_updated?: number | null;
  lessons_skipped?: number | null;
  conflicts_created?: number | null;
  error_message?: string | null;
  last_run?: string | null;
}

export interface ScheduleViewResponse {
  parse_status: ScheduleParseStatusResponse;
  conflicts: ScheduleConflict[];
  lessons: LessonResponse[];
  grouped_lectures: GroupedLectureResponse[];
  last_updated: string;
}
